main.tf

```diff
@@ -1,3 +1,46 @@
+locals {
+  egress_rules = flatten([
+    for key, rule in var.egress_rules : flatten([
+      for type, cidrs in {
+        "ipv4" = rule.cidr_ipv4,
+        "ipv6" = rule.cidr_ipv6
+        } : cidrs != null ? [
+        for cidr in cidrs : {
+          key                          = "${key}-${cidr}"
+          description                  = rule.description
+          from_port                    = rule.from_port
+          ip_protocol                  = rule.ip_protocol
+          to_port                      = rule.to_port
+          cidr_ipv4                    = type == "ipv4" ? cidr : null
+          cidr_ipv6                    = type == "ipv6" ? cidr : null
+          prefix_list_id               = rule.prefix_list_id
+          referenced_security_group_id = rule.referenced_security_group_id
+        }
+      ] : []
+    ])
+  ])
```

Comment:
This can be simplified to `{ "${key}-${cidr}" => rule }` or something to that effect. Once it's a map, the for_each in the resources will also be easier.